Sources/MongoSwift/BSON/Document.swift

```diff
@@ -181,26 +186,44 @@ public struct Document: ExpressibleByDictionaryLiteral, ExpressibleByArrayLitera
      *  ```
      */
     public subscript(key: String) -> BsonValue? {
-        get {
-            return DocumentIterator(forDocument: self, advancedTo: key)?.currentValue
-        }
-
+        get { return DocumentIterator(forDocument: self, advancedTo: key)?.currentValue }
         set(newValue) {
-            self.copyStorageIfRequired()
+            do {
+                try self.setValue(forKey: key, to: newValue)
+            } catch {
+                preconditionFailure("Failed to set the value for key \(key) to \(newValue ?? "nil"): \(error)")
+            }
+        }
+    }
 
-            guard let value = newValue else {
-                if !bson_append_null(self.data, key, Int32(key.count)) {
-                    preconditionFailure("Failed to set the value for key \(key) to null")
+    private mutating func setValue(forKey key: String, to newValue: BsonValue?) throws {
+        // if the key already exists in the `Document`, we need to replace it
+        if let existingType = DocumentIterator(forDocument: self, advancedTo: key)?.currentType {
+            // if the new type is the same and it's a fixed length type, we can overwrite
+            if let ov = newValue as? Overwritable, ov.bsonType == existingType {
+                self.copyStorageIfRequired()
+                // make a new iterator referencing our new storage. we already know the key is present
+                /// so initialization will succeed and ! is safe.
+                DocumentIterator(forDocument: self, advancedTo: key)!.overwriteCurrentValue(with: ov)
+
+            // otherwise, we just create a new document and replace this key
+            } else {
+                var newSelf = Document()
+                self.forEach {
+                    newSelf[$0.key] = $0.key == key ? newValue : $0.value
```

Comment:
This looks like it might have trouble with duplicate keys. Would it make sense to add some logic to guarantee that we only replace the first occurrence to be consistent with the Overwritable case above?

Reply:
sure, done. I think that as part of document validation, we should include adding our own validation against duplicate keys - libbson validation does not check for this 😱 - and then we can revisit what we are doing here